src/components/structures/MatrixChat.js

```diff
@@ -1203,21 +1204,44 @@ module.exports = React.createClass({
         } else if (screen.indexOf('user/') == 0) {
             const userId = screen.substring(5);
 
-            if (params.action === 'chat') {
-                this._chatCreateOrReuse(userId);
-                return;
-            }
+            // Wait for the first sync so that `getRoom` gives us a room object if it's
+            // in the sync response
+            const waitFor = this.firstSyncPromise ?
+                this.firstSyncPromise.promise : Promise.resolve();
+            waitFor.then(() => {
+                if (params.action === 'chat') {
+                    this._chatCreateOrReuse(userId);
+                    return;
+                }
+
+                // Get the member object for the current room, if a current room is set or
+                // we have a last_room in localStorage. The user might not be a member of
+                // this room (in which case member will be falsey).
+                let member;
+                const roomId = this.state.currentRoomId || localStorage.getItem('mx_last_room_id');
+                if (roomId) {
+                    const room = MatrixClientPeg.get().getRoom(roomId);
+                    if (room) {
+                        member = room.getMember(userId);
+                    }
+                }
 
-            this.setState({ viewUserId: userId });
-            this._setPage(PageTypes.UserView);
-            this.notifyNewScreen('user/' + userId);
-            const member = new Matrix.RoomMember(null, userId);
-            if (member) {
+                if (member) {
+                    // This user is a member of this room, so view the room
+                    dis.dispatch({
+                        action: 'view_room',
+                        room_id: roomId,
+                    });
+                } else {
+                    // This user is not a member of this room, show the user view
+                    member = new Matrix.RoomMember(roomId, userId);
```

Comment:
roomId should be null here since we're showing the user view, independent of a room (because they are not in the room)?